README
tlsprobe - a TLS/SSL testing utility (by Marco Bellaccini - marco.bellaccini[at!]gmail.com)

tlsprobe is a simple program that tests a server for TLS/SSL Cipher Suites support (client mode) or, listening for incoming TLS/SSL connections, discovers Cipher Suites offered by a client (server mode).
tlsprobe is Free Software, released under the Apache License, Version 2.0.

Website: http://marco-bellaccini.bitbucket.org/tlsprobe/
Bitbucket project page: https://bitbucket.org/marco-bellaccini/tlsprobe

In client mode, tlsprobe just sends ClientHello messages, and, by looking at server's reply (if any), determines whether the selected Cipher Suite(s) is supported or not.
tlsprobe builds TLS/SSL messages in a "byte-by-byte" fashion and so does not need any external library (i.e.: no OpenSSL, no GnuTLS, no libNSS).
The program supports both single Cipher Suite support check and a full-scan mode (in which it checks for support of all the cipher suites listed in the IANA TLS Cipher Suite Registry - http://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-4 ).
In server mode, it waits for a ClientHello to arrive and parse it, discovering offered Cipher Suites.

tlsprobe provides you with a visual feedback of the security/compatibility level of the various TLS/SSL Cipher Suites, based on Mozilla Server-Side-TLS Recommendations: https://wiki.mozilla.org/Security/Server_Side_TLS .
It highlights the discovered Cipher Suites with different colors, depending on the estimated security and compatibility level.

Working at TLS/SSL level, tlsprobe is generic and also supports probing non-https services (you just have to specify the right port for the higher-level protocol with the -p option).

tlsprobe supports SSL 3.0, TLS 1.0, TLS 1.1 and TLS 1.2.

Here follows some usage examples.

>>Client mode

List all Cipher Suites supported by www.google.com through TLS and SSL protocols (https, tcp port 443):

tlsprobe -F www.google.com

List all TLS Cipher Suites (i.e.: skip SSL3 scanning phase) supported by www.google.com (https, tcp port 443):

tlsprobe -j -F www.google.com

Tests if www.google.com (https, tcp port 443) supports cipher suite TLS_RSA_WITH_AES_128_CBC_SHA:

tlsprobe -c TLS_RSA_WITH_AES_128_CBC_SHA www.google.com

Tests if www.google.com (https, tcp port 443) supports cipher suite SSL_RSA_WITH_AES_128_CBC_SHA (with SSL 3.0 protocol):

tlsprobe -c SSL_RSA_WITH_AES_128_CBC_SHA www.google.com

List all Cipher Suites supported by Active Directory Domain Controller 10.0.0.5 with LDAP over ssl/tls (ldaps, tcp port 636):

tlsprobe -F -p 636 10.0.0.5

>>Server mode

Listen for TLS connections on TCP port 443 and print Cipher Suites offered by the client:

tlsprobe -S 0.0.0.0

Listen for TLS connections on TCP port 8443 and print Cipher Suites offered by the client:

tlsprobe -S -p 8443 0.0.0.0

For more help about possible options, try:

tlsprobe --help




ChangeLog
2016-10-04  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* retries 3 times when socket get stuck

2016-07-23  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* try to connect multiple times before giving up
	
2016-07-21  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* improved alert message support
	
2016-07-03  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* added Signature Algorithms Extension support in order to improve IIS/Schannel compatibility
	
2016-07-02  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* minor changes
	
2016-07-02  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* fixed some bugs related to server replies

2016-06-11  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* improved Cipher Suites evaluation

2016-06-11  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* realigned Cipher Suites evaluations with Mozilla specs

2016-06-11  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* improved ChaCha20-Poly1305 Cipher Suites support

2015-10-09  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* minor bugfix

2015-10-07  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* minor bugfix
	* added support for some old cipher suites

2015-10-06  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* fixed some threads-related bugs

2015-10-05  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* fixed some bugs

2015-10-04  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* added support for elliptic_curves and ec_point_formats TLS Extensions

2015-10-03  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* reorganized ClientHello generation procedure
	* added support for server_name TLS Extension
	* added ChaCha Cipher Suites support

2015-09-26  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* basic multi-thread support (enabled by default)
	* fixed nasty memory bug

2015-09-20  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* SSL3 support
	* TLS1.3 (draft) support
	* improved CS loading function
	* corrected TLS_EMPTY_RENEGOTIATION_INFO_SCSV and TLS_FALLBACK_SCSV handling

2015-09-13  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* Fixed cs_eval.dat
	
2015-09-13  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* Changed color codes
	* Quiet mode
	* Updated README

2015-09-12  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* Updated cs_eval.dat with one generated with last openssl version
	* Added unknown security category
	
2015-09-12  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* Added Cipher Suite Strength Evaluation with color feedback	
	* Fixed TIME_WAIT issue
	* Performed some code cleanup

2015-09-10  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* Added right AC_CHECK_HEADERS in configure.ac
	* Added server mode
	* Added TODO file

2015-09-06  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* Fixed Makefile.am

2015-09-06  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* Renamed project

2015-09-05  Marco Bellaccini <marco.bellaccini[at!]gmail.com>

	* First public version




src/arg.h
/* Copyright 2015 Marco Bellaccini (marco.bellaccini[at!]gmail.com)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#define ARGNUM 1


const char *argp_program_version =
  "tlsprobe 0.1";
const char *argp_program_bug_address =
  "<marco.bellaccini[at!]gmail.com>";

/* Program documentation. */
static char doc[] =
  "tlsprobe -- a TLS/SSL tester utility";

/* A description of the arguments we accept. */
static char args_doc[] = "host";

/* The options we understand. */
static struct argp_option options[] = {
  {"true-time",    'T', 0,      0,  "Insert real Unix timestamp in TLS messages instead of random data (see https://bugzilla.mozilla.org/show_bug.cgi?id=967923 for details)" },
  {"print-messages",    'm', 0,      0,  "Print sent/received ClientHello message" },
  {"port",   'p', "PORT", 0, "Set port to TCP port PORT (default is tcp/443 - HTTPS)" },
  {"tls-file",   'f', "FILE", 0, "Use IANA TLS Cipher Suites List file FILE (default is tls-parameters-4.csv located in /usr/local/share/tlsprobe)" },
  {"ssl-file",   'g', "FILE", 0, "Use SSL Cipher Suites List file FILE (default is ssl-parameters.csv located in /usr/local/share/tlsprobe)" },
  {"cs-eval-file",   'e', "FILE", 0, "Use Cipher Suites Evaluation file FILE (default is cs_eval.dat located in /usr/local/share/tlsprobe)" },
  {"cipher-suite",   'c', "CIPHER_SUITE_ID", 0, "CLIENT MODE SINGLE CIPHER SUITE PROBE: test if server supports cipher suite CIPHER_SUITE_ID (e.g. TLS_RSA_WITH_AES_128_CBC_SHA)" },
  {"full-scan",   'F', 0, 0, "CLIENT MODE FULL-SCAN: test the server for support of all the cipher suites listed in the IANA Cipher Suites List file" },
  {"timeout",   't', "TIMEOUT", 0, "Set the timeout at TIMEOUT [ms] for server reply.\n\
  Note: smaller timeouts mean faster scan, but may lead to unreliable results (i.e.: underestimation of the number of supported ciphers).\nDefault timeout is 500ms." },
  {"auto-timeout",   'a', 0, 0, "Autoset timeout by estimating RTT with ping (needs a shell with ping, tail, awk and cut utilities)." },
  {"tls-version",   'R', "VERSION", 0, "Use TLS version VERSION (default is VERSION=1.2)" },
  {"server-mode",   'S', 0, 0, "SERVER MODE: listen for incoming connections and list offered Cipher Suites when a ClientHello message is received" },
  {"quiet",   'q', 0, 0, "Be quiet, printing only the results (and a small subset of messages)." },
  {"skip-ssl",   'j', 0, 0, "Skip SSL3 scan when performing a full-scan." },
  {"disable-extensions",   'd', 0, 0, "Disable all TLS Extensions." },
  {"disable-server-name",   'n', 0, 0, "Disable TLS server_name Extension." },
  {"disable-sigalg",   's', 0, 0, "Disable TLS Signature Algorithms Extension." },
  {"disable-ec",   'x', 0, 0, "Disable TLS elliptic_curves Extension." },
  {"disable-ec-pf",   'y', 0, 0, "Disable TLS elliptic_curves_point_formats Extension." },
  {"thread-num",   'w', "TNUM", 0, "Set maximum number of threads to TNUM when performing a full-scan (default is 16)." },
  { 0 }
};

/* Used by main to communicate with parse_opt. */
struct arguments
{
  char *args[ARGNUM];                // host and port
  int truetime, port, printMessage, fullScanMode, timeout, autotimeout, cipherSuiteMode, serverMode, quiet, skipSSL, maxThreads, TLSExtensions, TLSSNExtension, TLSECExtension, TLSECPFExtension, TLSSAExtension;
  char *CS_file, *CS_eval_file, *CS_file_SSL;
  char *cipherSuite;
  char *tlsVer;
};



/* Parse a single option. */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  /* Get the input argument from argp_parse, which we
     know is a pointer to our arguments structure. */
  struct arguments *arguments = state->input;

  switch (key)
    {
    case 'T':
      arguments->truetime = 1;
      break;
	case 'm':
      arguments->printMessage = 1;
	  break;
	case 'p':
      arguments->port = atoi(arg);
	break;
	case 'f':
      arguments->CS_file = arg;
	break;
	case 'g':
      arguments->CS_file_SSL = arg;
	break;
	case 'e':
      arguments->CS_eval_file = arg;
	break;
	case 'c':
      arguments->cipherSuite = arg;
	  arguments->cipherSuiteMode = 1;
	break;
	case 'F':
      arguments->fullScanMode = 1;
	break;
	case 't':
      arguments->timeout = atoi(arg);
	break;
	case 'a':
      arguments->autotimeout = 1;
	break;
	case 'R':
      arguments->tlsVer = arg;
	break;
	case 'S':
      arguments->serverMode = 1;
	break;
	case 'q':
      arguments->quiet = 1;
	break;
	case 'j':
      arguments->skipSSL = 1;
	break;
	case 'd':
      arguments->TLSExtensions = 0;
	break;
	case 'n':
      arguments->TLSSNExtension = 0;
	break;
	case 'x':
      arguments->TLSECExtension = 0;
	break;
	case 'y':
      arguments->TLSECPFExtension = 0;
	break;
	case 's':
      arguments->TLSSAExtension = 0;
	break;
	case 'w':
      arguments->maxThreads = atoi(arg);
	break;

    case ARGP_KEY_ARG:
      if (state->arg_num >= ARGNUM)
        /* Too many arguments. */
        argp_usage (state);

      arguments->args[state->arg_num] = arg;

      break;

    case ARGP_KEY_END:
      if (state->arg_num < ARGNUM )
        /* Not enough arguments. */
        argp_usage (state);
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

/* Our argp parser. */
static struct argp argp = { options, parse_opt, args_doc, doc };


src/tls.h
/* Copyright 2015 Marco Bellaccini (marco.bellaccini[at!]gmail.com)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

/* This is mostly from TLS RFCs */

#include <inttypes.h>


//1-byte alignment for structures: inefficient but makes life easier for network stuff
#pragma pack(1)

typedef uint8_t uint8;
typedef uint8_t opaque;
typedef uint16_t uint16;
typedef struct {
	uint8_t msb;
	uint8_t nsb;
	uint8_t lsb;
} uint24;
typedef uint32_t uint32;

typedef struct {
       uint8 major;
       uint8 minor;
} ProtocolVersion;

ProtocolVersion version;
ProtocolVersion version30 = { 3, 0 };     // SSL 3.0
ProtocolVersion version10 = { 3, 1 };     // TLS 1.0
ProtocolVersion version11 = { 3, 2 };     // TLS 1.1
ProtocolVersion version12 = { 3, 3 };     // TLS 1.2

   
/* Alert Messages */


// AlertLevel

#define AL_WARNING 1
#define AL_FATAL 2

// AlertDescription

#define AD_CLOSE_NOTIFY 0
#define AD_UNEXPECTED_MESSAGE 10
#define AD_BAD_RECORD_MAC 20
#define AD_DECRYPTION_FAILED_RESERVED 21
#define AD_RECORD_OVERFLOW 22
#define AD_DECOMPRESSION_FAILURE 30
#define AD_HANDSHAKE_FAILURE 40
#define AD_NO_CERTIFICATE_RESERVED 41
#define AD_BAD_CERTIFICATE 42
#define AD_UNSUPPORTED_CERTIFICATE 43
#define AD_CERTIFICATE_REVOKED 44
#define AD_CERTIFICATE_EXPIRED 45
#define AD_CERTIFICATE_UNKNOWN 46
#define AD_ILLEGAL_PARAMETER 47
#define AD_UNKNOWN_CA 48
#define AD_ACCESS_DENIED 49
#define AD_DECODE_ERROR 50
#define AD_DECRYPT_ERROR 51
#define AD_EXPORT_RESTRICTION_RESERVED 60
#define AD_PROTOCOL_VERSION 70
#define AD_INSUFFICIENT_SECURITY 71
#define AD_INTERNAL_ERROR 80
#define AD_USER_CANCELED 90
#define AD_NO_RENEGOTIATION 100
#define AD_UNSUPPORTED_EXTENSION 110

typedef struct {
       uint8 level;
       uint8 description;
} Alert;
   
   
/* Hello Messages */
typedef struct { } HelloRequest;

typedef struct {
	uint32 gmt_unix_time;
	opaque random_bytes[28];
} Random;


typedef struct {
	uint8 a;
	uint8 b;
} CipherSuite;

typedef uint8_t CompressionMethod;

/* TLS Extensions for Hello Messages */

typedef struct {
	uint16 type;
	uint16 length;
} Extension;

// server_name extension data
typedef struct {
	uint16 list_length;
	uint16 name_type;
	uint8 name_length;
	char* name;
} ServerNameExData;

// elliptic curves extension data
typedef struct {
	uint16 ECLength;
	uint16* curves; // beware, you fixed 35 elements!!!
} ECExData;

// elliptic curves point formats extension data
typedef struct {
	uint8 formats_length;
	uint8* formats; // beware, you fixed 1 element!!!
} ECPFExData;

// TLS Extensions
#define TLS_EXT_SERVER_NAME 0x0000
#define TLS_EXT_SIGN_ALG 0x000d
#define TLS_EXT_EC 0x000a
#define TLS_EXT_EC_PF 0x000b

// server name
#define TLS_EXT_SERVER_NAME_HOSTNAME 0x0000

// signature algorithm hashes
#define TLS_EXT_SIGN_ALG_H_NONE 0
#define TLS_EXT_SIGN_ALG_H_MD5 1
#define TLS_EXT_SIGN_ALG_H_SHA1 2
#define TLS_EXT_SIGN_ALG_H_SHA224 3
#define TLS_EXT_SIGN_ALG_H_SHA256 4
#define TLS_EXT_SIGN_ALG_H_SHA384 5
#define TLS_EXT_SIGN_ALG_H_SHA512 6

// signature algorithms
#define TLS_EXT_SIGN_ALG_S_ANON 0
#define TLS_EXT_SIGN_ALG_S_RSA 1
#define TLS_EXT_SIGN_ALG_S_DSA 2
#define TLS_EXT_SIGN_ALG_S_ECDSA 3

// signature and hash alg ext
typedef struct {
	uint8 hash;
	uint8 signature;
} SignatureAndHashAlgorithm;

typedef struct {
	uint16 length;
	SignatureAndHashAlgorithm* data;
} SigExData;
      

// elliptic curves
/*
#define TLS_EXT_EC_SECT163K1 1
#define TLS_EXT_EC_SECT163R1 2
#define TLS_EXT_EC_SECT163R2 3
#define TLS_EXT_EC_SECT193R1 4
#define TLS_EXT_EC_SECT193R2 5
#define TLS_EXT_EC_SECT233K1 6
#define TLS_EXT_EC_SECT233R1 7
#define TLS_EXT_EC_SECT239K1 8
#define TLS_EXT_EC_SECT283K1 9
#define TLS_EXT_EC_SECT283R1 10
#define TLS_EXT_EC_SECT409K1 11
#define TLS_EXT_EC_SECT409R1 12
#define TLS_EXT_EC_SECT571K1 13
#define TLS_EXT_EC_SECT571R1 14
#define TLS_EXT_EC_SECP160K1 15
#define TLS_EXT_EC_SECP160R1 16
#define TLS_EXT_EC_SECP160R2 17
#define TLS_EXT_EC_SECP192K1 18
#define TLS_EXT_EC_SECP192R1 19
#define TLS_EXT_EC_SECP224K1 20
#define TLS_EXT_EC_SECP224R1 21
#define TLS_EXT_EC_SECP256K1 22
#define TLS_EXT_EC_SECP256R1 23
#define TLS_EXT_EC_SECP384R1 24
#define TLS_EXT_EC_SECP521R1 25
#define TLS_EXT_EC_BRAINPOOLP256R1 26
#define TLS_EXT_EC_BRAINPOOLP384R1 27
#define TLS_EXT_EC_BRAINPOOLP512R1 28
#define TLS_EXT_EC_FFDHE2048 256
#define TLS_EXT_EC_FFDHE3072 257
#define TLS_EXT_EC_FFDHE4096 258
#define TLS_EXT_EC_FFDHE6144 259
#define TLS_EXT_EC_FFDHE8192 260
#define TLS_EXT_EC_ARB_EXP_PRIME_CURV 65281
#define TLS_EXT_EC_ARB_EXP_CHAR2_CURV 65282
*/


// elliptic curves point formats
#define TLS_EXT_EC_PF_UN 0
   

typedef struct {
	ProtocolVersion client_version;
	Random random;
	uint8 session_id;
	uint16 cipher_suites_length;
	CipherSuite* cipher_suites;
	uint8 compression_methods_length;
	CompressionMethod* compression_methods;
	uint16 extensions_length;
} ClientHello;


// Compression Methods

#define CM_NO_COMPRESSION 0


/* Handshake Protocol */
typedef uint8_t HandshakeType;
   
// HandshakeType

#define HT_HELLO_REQUEST 0
#define HT_CLIENT_HELLO 1
#define HT_SERVER_HELLO 2
#define HT_CERTIFICATE 11
#define HT_SERVER_KEY_EXCHANGE  12
#define HT_CERTIFICATE_REQUEST 13
#define HT_SERVER_HELLO_DONE 14
#define HT_CERTIFICATE_VERIFY 15
#define HT_CLIENT_KEY_EXCHANGE 16
#define HT_FINISHED 20

typedef struct {
	HandshakeType msg_type;
	uint24 length;
} HandshakeClientHello;
   
   
/* Record Layer */

// ContentType

#define CT_CHANGE_CIPHER_SPEC 20
#define CT_ALERT 21
#define CT_HANDSHAKE 22
#define CT_APPLICATION_DATA 23

typedef struct {
	uint8 type;
	ProtocolVersion version;
	uint16 length;
} TLSPlaintext;
   
typedef struct {
	uint8 type;
	ProtocolVersion version;
	uint16 length;
	Alert body;
} TLSPlaintextAL;
   
   
/* The Cipher Suite */

typedef struct {
	char name[100];
	CipherSuite id;
} CSuiteDesc;

typedef struct {
	CSuiteDesc* CSArray;
	int nol;
} CSuiteList;


/* Cipher Suite Evaluation */

typedef struct {
	CipherSuite *modern;
	CipherSuite *intermediate;
	CipherSuite *old;
	CipherSuite *all;
	int modern_size;
	int intermediate_size;
	int old_size;
	int all_size;
} CSuiteEvals;

   
   




